### ISSUE TITLE
Detsim in golang

### ISSUE THREAD

**Issue by vitorguidi** [eatonphil/raft-rs#2]
I noticed there is a seed for the testing in the screenshot for this repo, which belongs to a deterministic simulation, tiger beetle style. Having said that, is there any reason why the same approach would not be possible in golang?

Asking because I intend to implement raft, but I do not want to learn rust in order to do that. Also, I do not see a detsim in your [previous go implementation](https://github.com/eatonphil/goraft), so this made me wonder if there is any language level restriction, in order to do it in go.

**Comment by eatonphil** [comment]
Yeah that was a sketch to see if it was feasible in Rust. The gist for Rust and Go is that either threads/goroutines themselves or synchronizing between them is nondeterministic. So you'd need to do the whole thing single-threaded which means you need some sort of async io. But neither Go nor Rust have builtin support for async io. Which would make an implementation without third-party libraries (which I prefer to do) pretty cumbersome.

**Comment by vitorguidi** [comment]
It might be a very dumb question, but I honestly do not understand why:

```So you'd need to do the whole thing single-threaded which means you need some sort of async io```

To be more precise, the jump from ```determinism``` to ```async io``` is not immediately clear to mel

Gonna throw my mental model here, on how async IO can help achieve determinism. Please help me understand the gist of it:

- Main program is called, we define an executor and a scheduler (from the java world). All of them are running on the same thread
- We launch servers and clients as async tasks (ie, raft server), and they get scheduled
- Main program (simulation) delegates control to the scheduler
- Scheduler gets first in priority queue, pops it, runs in executor
- Executor gladly runs raft client, async operations within the client are enqueued in the executor; Whenever we find an `await` statement, raft client yields control;
- Scheduler gets control back, pops from PQ, tosses to executor
- Repeat until done, insert invriant assertions in between

Is that how it works? Trying to form a mental model as I read here => https://jbaker.io/2022/05/09/project-loom-for-distributed-systems/
